### DISCUSSION TITLE
为什么数据库不应该使用外键 - 面向信仰编程 · /whys-the-design-database-foreign-key

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#619] [Announcements]
https://draveness.me/whys-the-design-database-foreign-key/

**Comment by zeromake** [comment]
话说现在接触到的系统里都很少直接删除行数据了，都是用 `deleted` 或者 `deleted_at` 做归档化。

**Comment by draveness** [comment]
> 话说现在接触到的系统里都很少直接删除行数据了，都是用 `deleted` 或者 `deleted_at` 做归档化。

这确实也是一种常见的做法

**Comment by mrdear** [comment]
上大学时，学习数据库各种知识，三范式，触发器，存储过程，视图等，但到工作后发现很少使用到这些特性，业务中关注的只是存储数量级以及索引，把DB当成一个纯存储组件，只要能保存数据，并且提供快速查询足够了。那既然是这样，为什么没有出现专门为互联网一些业务设计的DB呢，只考虑基本的存储，索引，事务，也有点好奇这一点

**Reply by wuzhipeng1991** [reply]
mongodb啊

**Comment by draveness** [comment]
> 上大学时，学习数据库各种知识，三范式，触发器，存储过程，视图等，但到工作后发现很少使用到这些特性，业务中关注的只是存储数量级以及索引，把DB当成一个纯存储组件，只要能保存数据，并且提供快速查询足够了。那既然是这样，为什么没有出现专门为互联网一些业务设计的DB呢，只考虑基本的存储，索引，事务，也有点好奇这一点

时代变了，很多时候我们也不需要这么复杂的抽象，很多非关系型数据库都可以满足这三个需求，这可能也是 NoSQL 目前应用越来越广的原因吧。

**Comment by tottea** [comment]
大佬了解TiDB吗，期待能讲一期分布式关系型数据库相关的

**Comment by draveness** [comment]
> 大佬了解TiDB吗，期待能讲一期分布式关系型数据库相关的

TiDB 不太了解，不过可以讲讲分布式 SQL

**Comment by aiyogit** [comment]
请问作者的配图都是如何制作的？

**Comment by zeromake** [comment]
> @aiyogit
> 请问作者的配图都是如何制作的？

又一个不好好看文章的兄弟，文章的最后是什么

**Comment by aiyogit** [comment]
> @zeromake
> > @aiyogit
> > 请问作者的配图都是如何制作的？
> 
> 又一个不好好看文章的兄弟，文章的最后是什么

现在看到了，谢谢。

**Comment by scbizu** [comment]
唔，benchmark是不是有点问题 ？ Baseline的代码应该增加额外的SELECT操作吧 。

**Comment by draveness** [comment]
> 唔，benchmark是不是有点问题 ？ Baseline的代码应该增加额外的SELECT操作吧 。

我觉没啥问题，这里看的是外键带来多少额外开销，不是模拟外键功能与外键的开销对比

**Comment by siriuscc** [comment]
对于python gevent 和asyncio, 还有go 的协程，感觉还处于在用却看不到本质，不知道 博主是否有一些见解可以分享。

**Comment by draveness** [comment]
> 对于python gevent 和asyncio, 还有go 的协程，感觉还处于在用却看不到本质，不知道 博主是否有一些见解可以分享。

Go 语言的调度器和 Goroutine 可以看这个 https://draveness.me/golang/docs/part3-runtime/ch06-concurrency/golang-goroutine/

**Comment by flonly2008** [comment]
请教下博主，我也曾经想多次写blog来整理自己的思路和大家交流，后来发现时间实在不够就放弃了。因为每天下班就23点多了。请教博主的时间是怎么安排的

**Comment by draveness** [comment]
> 请教下博主，我也曾经想多次写blog来整理自己的思路和大家交流，后来发现时间实在不够就放弃了。因为每天下班就23点多了。请教博主的时间是怎么安排的

来看看这个 https://draveness.me/few-words-time-management/

**Comment by Evan6998** [comment]
感谢作者的这些文章

**Comment by towavephone** [comment]
老哥真强，这些文章看的我一愣一愣的

**Comment by Jarvan-via** [comment]
我愿意称老哥为最强

**Comment by JIAQIA** [comment]
> @mrdear
> 上大学时，学习数据库各种知识，三范式，触发器，存储过程，视图等，但到工作后发现很少使用到这些特性，业务中关注的只是存储数量级以及索引，把DB当成一个纯存储组件，只要能保存数据，并且提供快速查询足够了。那既然是这样，为什么没有出现专门为互联网一些业务设计的DB呢，只考虑基本的存储，索引，事务，也有点好奇这一点

这不就是NoSQL数据了么？但NoSQL很难保证事务，事务需要自己保证。

**Comment by zeromake** [comment]
@youncccat 
> 请问一下文章里的图是用什么工具生成的呢？

不好好看文章+1，文章结尾最后是什么……

**Comment by AustinWp** [comment]
好文章，最后的问题也是我好奇的

**Comment by azture** [comment]
请问博主中文是什么字体😁

**Comment by zhentao-huang** [comment]
这种做法太片面。任何功能特性都有可能有好有坏。按作者说法，直接把这功能在RDBMS中删除就行了，为什么会广泛存在？
从一般使用者角度来看，外键是一种纯纯的自我约束。删除它的定义，不会对表的使用产生任何阻碍，反而更加自由。从数据库管理员看来，这个外键带来的是吞吐量的降低，和各种运维的掣肘。这些都可以理解的。
了解使用这些代价，让我们回过头来看看，外键带来的真实好处。
大型项目，或大型项目中的子项目，都面临牵涉者众多的问题，这其中的理念的不一致会引起很大的麻烦。从数据库层面看，外键其实很好的表达了一种合约，对自己的约束，但可获得安全可靠和防误用。减少沟通成本。
从软件设计层面，数据库的提供了很好的数据抽象。大型项目，如果只采用裸数据到祼结构库这样的方式存贮数据。只会带来适用性差，扩展性差。解决这些问题，又可能采用预留字段，重复数据落表等方法。进而带来，数据库庞大冗余，性能下降，越来越难以维护，成本无法负担等问题，最后积重难返。这个问题在比较初级的开发水平很容易发生，但也确实是普便情况。好的数据抽象，能有效减缓，解决这些问题，而这又离不开RDBMS提供的一些看上去不太实用的如外键之类的功能。简单的说，还是大家不太会用。如果因噎废食，断然否定，其实是得不偿失的。
本文的角度，可能是从更利于运维的角度来看这个问题，但其实，比如mysql，也提供了如 `foreign_key_checks`这样的变量，可以提供运维方便的手段，以处理无须理解业务，而安全的操作。多了解一些RDBMS这些方面的设计，了解问题的代价，积极寻求解决方案，可更好的提高运维水平。即使前水平低下，也不能一概而论的就直接否定的这个功能。等以后开拓了眼界，相信你会了解这些设计的精妙。好看又实用的Architecture，都由是依赖于一些精巧的结构的。

**Reply by draveness** [reply]
你看完了么，要不然再读读？

> 外键提供的几种在更新和删除时的不同行为都可以帮助我们保证数据库中数据的一致性和引用合法性，但是外键的使用也需要数据库承担额外的开销，在大多数服务都可以水平扩容的今天，高并发场景中使用外键确实会影响服务的吞吐量上限。在数据库之外手动实现外键的功能是可能的，但是却会带来很多维护上的成本或者需要我们在数据一致性上做出一些妥协。

> 当我们考虑应不应该在数据库中使用外键时，需要关注的核心我们的数据库承担这部分计算任务后会不会影响系统的可用性，在使用时也不应该一刀切的决定用或者不用外键，应该根据具体的场景做决策

**Reply by zhentao-huang** [reply]
这部分有用吗？你看看你的标题，引用的人不会看这么仔细，直接引用标题的结论。你说的回转的话和就和｀本品不能替代药品｀一样。当得了肺癌，才回过头来认识来这｀吸烟有害健康｀很重要一样。
我并不否定你的一些论据，但是你要了解你说的最后呈现给人的观感是什么。比如标题改成讨论外键的一些问题，是不是更合适？
你看到这问题，我猜你大概率会直接采用这种切断功能的方法。我的看法是承担，减轻，解决问题，而绝对不限制功能。初级开发者有可能会试用这功能，也可能最后放弃，这都没问题。但如我所见一些DBA，拿一些网上如你标题这般的观点作限制，最后误导开发，这就不好了。
不然RDBMS有演进的必要吗？Trigger, Function, procedure都有复杂化系统的危险，为什么不都砍了，大家都去用foxbase foxpro好了。

**Reply by draveness** [reply]
你说的很有道理，希望可以启发更多的人👍

**Comment by haipiaodev** [comment]
不使用foreign key之后，额外的select for update带来的开销不用考虑了吗？如果有两种方案的性能对比你的结论才完整。
